tests/thandle_ut/thandle_ut.c

```diff
@@ -0,0 +1,407 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifdef __cplusplus
+#include <cstdlib>
+#include <cstddef>
+#else
+#include <stdlib.h>
+#include <stddef.h>
+#endif
+
+#include "azure_macro_utils/macro_utils.h"
+#include "testrunnerswitcher.h"
+
+void* my_gballoc_malloc(size_t size)
+{
+    return malloc(size);
+}
+
+void my_gballoc_free(void* ptr)
+{
+    free(ptr);
+}
+
+#define ENABLE_MOCKS
+#include "umock_c/umock_c.h"
+#include "azure_c_util/gballoc.h"
+
+#undef ENABLE_MOCKS
+
+#include "azure_c_util/thandle.h"
+#include "thandle_user.h"
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+#define TEST_A_DEFINE 1
+static const int TEST_A = TEST_A_DEFINE;
+
+#define TEST_S_DEFINE "a"
+static const char* TEST_S = TEST_S_DEFINE;
+
+#define TEST_S2_DEFINE "aa"
+static const char* TEST_S2 = TEST_S2_DEFINE;
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+{
+    ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
+}
+
+/*the following data structures etc are needed because THANDLE_FREE cannot be called with NULL from a THANDLER(T) holder*/
+#define A_B_FIELDS        \
+    int, a,                 \
+    int, b
+
+MU_DEFINE_STRUCT(A_B, A_B_FIELDS);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+    
+    THANDLE_TYPE_DECLARE(A_B);
+    THANDLE_TYPE_DEFINE(A_B); 
+
+#ifdef __cplusplus
+    }
+#endif
+
+BEGIN_TEST_SUITE(thandle_unittests)
+
+TEST_SUITE_INITIALIZE(setsBufferTempSize)
+{
+    g_testByTest = TEST_MUTEX_CREATE();
+    ASSERT_IS_NOT_NULL(g_testByTest);
+
+    umock_c_init(on_umock_c_error);
+
+    REGISTER_GLOBAL_MOCK_HOOK(gballoc_malloc, my_gballoc_malloc);
+    REGISTER_GLOBAL_MOCK_HOOK(gballoc_free, my_gballoc_free);
+}
+
+TEST_SUITE_CLEANUP(TestClassCleanup)
+{
+    umock_c_deinit();
+
+    TEST_MUTEX_DESTROY(g_testByTest);
+}
+
+TEST_FUNCTION_INITIALIZE(f)
+{
+    if (TEST_MUTEX_ACQUIRE(g_testByTest))
+    {
+        ASSERT_FAIL("our mutex is ABANDONED. Failure in test framework");
+    }
+
+    umock_c_reset_all_calls();
+}
+
+TEST_FUNCTION_CLEANUP(cleans)
+{
+    TEST_MUTEX_RELEASE(g_testByTest);
+}
+
+/*Tests_SRS_THANDLE_02_013: [ THANDLE_MALLOC shall allocate memory. ]*/
+/*Tests_SRS_THANDLE_02_014: [ THANDLE_MALLOC shall initialize the reference count to 1, store dispose and return a T* . ]*/
```

Comment:
Could we have that region-like comment with /* THANDLE_MALLOC */. I just found that useful a couple of times in my latest changes and it's not a lot to add them to ask for :-).

Reply:
I do not like "regions" because I believe they are redundant with the rest of the text in the file.

1. redundant with the TEST_FUNCTIOIN(API_NAME_does_something)
2. redundant with the /*TESTS_SRS_BALBALA_... where it explains what it does

I will add them anyway.


---
In reply to: [385375049](https://github.com/Azure/azure-c-util/pull/16#discussion_r385375049) [](ancestors = 385375049)